src/core/sio.h

```diff
@@ -162,29 +72,190 @@ class SIO {
+            RESET = 0x0040,
+            RX_IRQMODE = 0x0100,  // FIFO byte count, to-do: implement
+            TX_IRQEN = 0x0400,
+            RX_IRQEN = 0x0800,
+            ACK_IRQEN = 0x1000,
+            WHICH_PORT = 0x2000,  // 0=/JOY1, 1=/JOY2
+        };
+    };
+    enum {
+        // MCD flags
+        MCDST_CHANGED = 0x08,
+    };
+
+    struct PAD_Commands {
+        enum : uint8_t {
+            Read = 0x42,  // Read Command
+            None = 0x00,  // No command, idle state
+            Error = 0xFF  // Bad command
+        };
+    };
+    struct DeviceType {
+        enum : uint8_t {
+            None = 0x00,        // No device selected yet
+            PAD = 0x01,         // Pad Select
+            NetYaroze = 0x21,   // Net Yaroze Select
+            MemoryCard = 0x81,  // Memory Card Select
+            Ignore = 0xFF,      // Ignore incoming commands
+        };
+    };
+    struct SelectedPort {
+        enum : uint16_t {
+            Port1 = 0x0000,
+            Port2 = 0x2000,
+        };
+    };
+
+    template <size_t buffer_size, typename T>
+    class FIFO {
+      public:
+        ~FIFO() { clear(); }
+
+        void clear() {
+            while (!queue_.empty()) queue_.pop();
+        }
+        bool isEmpty() { return queue_.empty(); }
+        T peek() {
+            T ret = T();
+            if (!queue_.empty()) {
+                ret = queue_.front();
+            }
+
+            return ret;
+        }
+        T pull() {
+            T ret = T();
+            if (!queue_.empty()) {
+                ret = queue_.front();
+                queue_.pop();
+            }
+
+            return ret;
+        }
+        void push(T data) {
+            if (queue_.size() >= buffer_size) {
+                queue_.back() = data;
+            } else {
+                queue_.push(data);
+            }
+        }
+
+        size_t size() { return queue_.size(); }
+
+      private:
+        std::queue<T> queue_;
+    };
+
+    friend MemoryCard;
     friend SaveStates::SaveState SaveStates::constructSaveState();
+
+    static const size_t s_padBufferSize = 0x1010;
+
+    bool isReceiveIRQReady();
+    bool isTransmitReady();
+    inline void scheduleInterrupt(uint32_t eCycle) {
+        g_emulator->m_cpu->scheduleInterrupt(PSXINT_SIO, eCycle);
+#if 0
+// Breaks Twisted Metal 2 intro
+        m_statusReg &= ~RX_FIFONOTEMPTY;
+        m_statusReg &= ~TX_DATACLEAR;
+#endif
+    }
+    void transmitData();
+    void updateFIFOStatus();
+    void writePad(uint8_t value);
+
+    SIORegisters m_regs = {
+        .status = StatusFlags::TX_DATACLEAR | StatusFlags::TX_FINISHED,  // Transfer Ready and the Buffer is Empty
+    };
+
+    uint8_t m_currentDevice = DeviceType::None;
+
+    // Pads
+    uint8_t m_buffer[s_padBufferSize];
+    uint32_t m_bufferIndex;
+    uint32_t m_maxBufferIndex;
+    uint32_t m_padState;
+
+    MemoryCard m_memoryCard[s_cardCount] = {this, this};
+
+    FIFO<8, uint8_t> m_rxFIFO;
```

Comment:
Purely for cosmetic reasons, I would've swapped the template parameters here.

Reply:
Makes sense